src/Nethermind/Nethermind.Evm/Witness/VerkleExecWitness.cs

```diff
@@ -237,53 +234,62 @@ public byte[][] GetAccessedKeys()
     {
-        return AccessAccountSubTree(address, UInt256.Zero, AccountHeader.CodeSize, isWrite);
+        return AccessAccountSubTree(address, UInt256.Zero, AccountHeader.CodeSize, ref gasAvailable, isWrite);
     }
 
-    private long AccessCodeHash(Address address, bool isWrite = false)
+    private bool AccessCodeHash(Address address, ref long gasAvailable, bool isWrite = false)
     {
-        return AccessAccountSubTree(address, UInt256.Zero, AccountHeader.CodeHash, isWrite);
+        return AccessAccountSubTree(address, UInt256.Zero, AccountHeader.CodeHash, ref gasAvailable, isWrite);
     }
 
-    private long AccessAccountSubTree(Address address, UInt256 treeIndex, byte subIndex, bool isWrite = false)
+    private bool AccessAccountSubTree(Address address, UInt256 treeIndex, byte subIndex, ref long gasAvailable, bool isWrite = false)
     {
-        if (address.IsPrecompile(Osaka.Instance)) return 0;
-        return AccessKey(AccountHeader.GetTreeKey(address.Bytes, treeIndex, subIndex), isWrite);
+        if (address.IsPrecompile(Osaka.Instance)) return true;
+        return AccessKey(AccountHeader.GetTreeKey(address.Bytes, treeIndex, subIndex), ref gasAvailable, isWrite);
     }
 
-    private long AccessKey(Hash256 key, bool isWrite = false, bool leafExist = false)
+    private bool AccessKey(Hash256 key, ref long gasAvailable, bool isWrite = false, bool leafExist = false)
     {
-        long accessCost = 0;
-        if (_accessedLeaves.Add(key)) accessCost += GasCostOf.WitnessChunkRead;
-        if (_accessedSubtrees.Add(key.Bytes[..31].ToArray())) accessCost += GasCostOf.WitnessBranchRead;
+        long requiredGas = 0;
+        if (!_accessedLeaves.Contains(key)) requiredGas += GasCostOf.WitnessChunkRead;
+        if (!_accessedSubtrees.Contains(key.Bytes[..31].ToArray())) requiredGas += GasCostOf.WitnessBranchRead;
 
-        if (isWrite)
-        {
-            if (_modifiedLeaves.Add(key)) accessCost += GasCostOf.WitnessChunkWrite;
-            if (_modifiedSubtrees.Add(key.Bytes[..31].ToArray())) accessCost += GasCostOf.WitnessBranchWrite;
+        if (requiredGas > gasAvailable) return false;
+        gasAvailable -= requiredGas;
 
-            if (!_verkleWorldState.ValuePresentInTree(key))
-            {
-                accessCost += GasCostOf.WitnessChunkFill;
-            }
-        }
-        return accessCost;
+        if (!_accessedLeaves.Contains(key)) _accessedLeaves.Add(key);
+        if (!_accessedSubtrees.Contains(key.Bytes[..31].ToArray())) _accessedSubtrees.Add(key.Bytes[..31].ToArray());
+
+        if (!isWrite) return true;
+
+        requiredGas = 0;
+        if (!_modifiedLeaves.Contains(key)) requiredGas += GasCostOf.WitnessChunkWrite;
+        if (!_modifiedSubtrees.Contains(key.Bytes[..31].ToArray())) requiredGas += GasCostOf.WitnessBranchWrite;
+        if (!_verkleWorldState.ValuePresentInTree(key)) requiredGas += GasCostOf.WitnessChunkFill;
```

Comment:
```suggestion
        if (!_modifiedSubtrees.Contains(key.Bytes[..31].ToArray())) 
        {
                requiredGas += GasCostOf.WitnessBranchWrite;
                if (!_verkleWorldState.ValuePresentInTree(key)) requiredGas += GasCostOf.WitnessChunkFill;
        }
```

since in the same transaction, if we try to charge for the witness again - we will charge WitnessChunkFill everytime as we dont push the changes to the tree regularly